### PR TITLE
[mono] Fix warnings on OSX, re-enable -Werror on CI with CMake

### PR DESCRIPTION
**Pull request by CoffeeFlux**
Some warnings snuck in after we lost this in the CMake transition. Contributes to https://github.com/dotnet/runtime/issues/41357

### CODE REVIEW

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/sgen/sgen-descriptor.c [hunk 1]
```diff
@@ -129,7 +129,7 @@ mono_gc_make_descr_for_object (gsize *bitmap, int numbits, size_t obj_size)
 	}
 
 	if (first_set < 0) {
-		SGEN_LOG (6, "Ptrfree descriptor %p, size: %ld", (void*)desc, (long)stored_size);
+		SGEN_LOG (6, "Ptrfree descriptor %llu, size: %zu", (guint64)desc, stored_size);
```
Comment (lambdageek) [inline src/mono/mono/sgen/sgen-descriptor.c:132]:
```suggestion
		SGEN_LOG (6, "Ptrfree descriptor %" PRIu64 ", size: %zu", (uint64_t)desc, stored_size);
```
`"%llu"` isn't portable

File: src/mono/mono/sgen/sgen-descriptor.c [hunk 2]
```diff
@@ -142,7 +142,7 @@ mono_gc_make_descr_for_object (gsize *bitmap, int numbits, size_t obj_size)
 	/* we know the 2-word header is ptr-free */
 	if (last_set < BITMAP_NUM_BITS + OBJECT_HEADER_WORDS && stored_size <= SGEN_MAX_SMALL_OBJ_SIZE) {
 		desc = DESC_TYPE_BITMAP | ((*bitmap >> OBJECT_HEADER_WORDS) << LOW_TYPE_BITS);
-		SGEN_LOG (6, "Largebitmap descriptor %p, size: %ld, last set: %d", (void*)desc, (long)stored_size, last_set);
+		SGEN_LOG (6, "Largebitmap descriptor %llu, size: %zu, last set: %d", (guint64)desc, stored_size, last_set);
```
Comment (lambdageek) [inline src/mono/mono/sgen/sgen-descriptor.c:145]:
```suggestion
		SGEN_LOG (6, "Largebitmap descriptor %" PRIu64 ", size: %zu, last set: %d", (uint64_t)desc, stored_size, last_set);
```

File: src/mono/mono/sgen/sgen-descriptor.c [hunk 3]
```diff
@@ -153,7 +153,7 @@ mono_gc_make_descr_for_object (gsize *bitmap, int numbits, size_t obj_size)
 		 */
 		if (first_set < 256 && num_set < 256 && (first_set + num_set == last_set + 1)) {
 			desc = DESC_TYPE_RUN_LENGTH | stored_size | (first_set << 16) | (num_set << 24);
-			SGEN_LOG (6, "Runlen descriptor %p, size: %ld, first set: %d, num set: %d", (void*)desc, (long)stored_size, first_set, num_set);
+			SGEN_LOG (6, "Runlen descriptor %llu, size: %zu, first set: %d, num set: %d", (guint64)desc, stored_size, first_set, num_set);
```
Comment (lambdageek) [inline src/mono/mono/sgen/sgen-descriptor.c:156]:
```suggestion
			SGEN_LOG (6, "Runlen descriptor %" PRIu64 ", size: %zu, first set: %d, num set: %d", (uint64_t)desc, stored_size, first_set, num_set);
```

File: src/mono/mono.proj [hunk 4]
```diff
@@ -60,6 +60,11 @@
     <Error Condition="'$(TargetsAndroid)' == 'true' and '$(ANDROID_NDK_ROOT)' == ''" Text="Error: You need to set the ANDROID_NDK_ROOT environment variable pointing to the Android NDK root." />
   </Target>
 
+  <!-- CI specific build options -->    
+  <ItemGroup Condition="'$(ContinuousIntegrationBuild)' == 'true' and ('$(TargetsOSX)' == 'true' or '$(TargetsBrowser)' == 'true')">
```
Comment (lambdageek) [inline src/mono/mono.proj:64]: I'm not a huge fan of making the CI build do something without giving an easy way to do the same thing in a local build.

tl;dr - we should have a property that you can pass to `build.sh` to treat warnings as errors. (Maybe there's already a way to pass `-DENABLE_WERROR=1` - I'm not sure, I didn't check)

Comment (CoffeeFlux) [reply src/mono/mono.proj:64]: You should see the exact warnings in a local build. I'm not changing the warnings we use, just whether they're going to make the build fail or not, so I don't think it's a big deal. I'm not sure how easy it is to thread `-DENABLE_WERROR` down from `build.sh`, though it's easy when you're invoking CMake directly.

File: src/mono/mono/sgen/sgen-descriptor.c (hunk 1 above)
Comment (CoffeeFlux) [reply src/mono/mono/sgen/sgen-descriptor.c:132]: It's technically undefined, but isn't it fine everywhere with C99? Happy to fix, but I don't think it actually matters.

Comment (lambdageek) [reply src/mono/mono/sgen/sgen-descriptor.c:132]: I don't think so.  Pretty sure it won't work on msvc.

File: src/mono/mono.proj (hunk 4 above)
Comment (lambdageek) [reply src/mono/mono.proj:64]: > I'm not sure how easy it is to thread `-DENABLE_WERROR` down from `build.sh`, though it's easy when you're invoking CMake directly.

is the workflow of invoking cmake directly documented somewhere?  IMO we should make it as easy as possible for devs to fix warnings-as-errors problems locally.

Comment (CoffeeFlux) [reply src/mono/mono.proj:64]: Not sure, but again you can just fix the warnings; the issues reported are the same, it's just a question of whether they block the build. I don't think it's worth holding this PR up for that and we should open a separate issue.

Anyway, there was substantial resistance to making -Werror the default locally but I'd be happy to see a build.sh flag for it. I think CoreCLR has -Werror on CI, and I'm not sure how they handle this, but maybe such an addition would be mutually beneficial.

File: src/mono/mono/sgen/sgen-descriptor.c (hunk 1 above)
Comment (vargaz) [reply src/mono/mono/sgen/sgen-descriptor.c:132]: Its pretty hard to print out a guint64, using %p and GUINT_TO_POINTER () might fix it.

File: src/mono/mono.proj (hunk 4 above)
Comment (vargaz) [reply src/mono/mono.proj:64]: We can add an option to mono.proj if needed to pass more flags to cmake, but it's a separate issue.

File: src/mono/mono/sgen/sgen-descriptor.c (hunk 1 above)
Comment (lateralusX) [reply src/mono/mono/sgen/sgen-descriptor.c:132]: Under VS2019 %llu works:

printf ("Max uint64_t: %llu", UINT64_MAX);

Max uint64_t: 18446744073709551615

and PRIu64 is defined to llu on msvc as well.

#define PRIu64       "llu"

but I guess including inttypes.h and use %" PRIu64 " is the more portable alternative.

**Comment by CoffeeFlux** [comment]
Blocked on remaining eventpipe warnings.

**Comment by lateralusX** [comment]
eventpipe warnings and osx build should be fixed.

**Review by akoeplinger — APPROVED** [review]

**Comment by lateralusX** [comment]
@CoffeeFlux Any plan on resolving the last wasm issue? I will probably need the eventpipe change for another PR during the week.

**Comment by CoffeeFlux** [comment]
I'll handle wasm separately. The remaining issue looks weird and I don't have time to investigate right now.

**Comment by CoffeeFlux** [comment]
It looks like none of the failures are build-related other than Android, which is hung rather than erroring.
